### DISCUSSION TITLE
Externalize 3dTiles support from itowns?

### DISCUSSION THREAD

**Discussion by mgermerie** [iTowns/itowns#1917] [Next Features]
We are studying possibilities in terms of libraries to enhance 3dTiles support. At the moment, we are considering : 

- 3dTilesRendererJS 
- Three-Loader-3dTiles
- Loaders.gl

We shall write our findings here.

**Comment by mgermerie** [comment]
After testing three-loader-3dtiles integration in iTowns (on [this branch](https://github.com/mgermerie/itowns/tree/misc/three-loader-3dtiles), with [this example](https://github.com/mgermerie/itowns/blob/misc/three-loader-3dtiles/examples/test-planar.html) and [this one](https://github.com/mgermerie/itowns/blob/misc/three-loader-3dtiles/examples/test-globe.html)), here are a few points : 

- The version of Three.js supported by three-loader-3dtiles is `0.137.0` while iTowns currently comes with `0.140.2`. Adding a dependency to three-loader-3dtiles in iTowns would therefore require that we upgrade its Three.js version, and that we keep upgrading it as we need to upgrade iTowns Three.js version.
- The library, though easy to use, would need some entry points like callbacks to ease its integration, E.g :  it would need a callback when a mesh is added to the main `Group` (called `model` in the library) which would allow us to tweak meshes. What we do for now is traversing the main `Group` children. 
- We would also need to work on the update : what we did for now is calling three-loader-3dtiles model `update` method before each time iTowns calls the rendered. If all iTowns `Layers` are loaded and displayed, iTowns won't call the renderer even. All 3dtiles data which were not fully loaded then won't therefore be rendered.
- When tested over a large `b3dm` 3dTiles dataset, using three-loader-3dtiles proved to slower than iTowns native support. Tests on different datasets (point clouds, i3dm) would be needed for further validation of this point though.

**Comment by jailln** [comment]
Here is a summary of the tests of 3DTilesRendererJS integration in itowns (see this [branch](https://github.com/jailln/itowns/tree/poc/3dtiles-renderer-js)):
* Pros:
  * Cesium ion support
  * Some work already done on the i3dm support (but our i3dm tilesets are not displayed)
  * clean architecture
  * less maintenance in itowns
* Cons: 
  * Performances are not as good as with itowns: loading and rendering time can sometimes be 2x slower than itowns.
  * Loss of viewer request volume support
  * Loss of gltf 1.0 support (and of old 3d tiles tilesets specific gltf attributes)
  * Tiles loading/refinement doesn't seem coherent sometimes 
  * Important bus factor (only one dev)
  * 3d tiles next implementation is not guaranteed
  * not guaranteed to have the possibility to manage the cache as precisely as in itowns

**Comment by jailln** [comment]
The tests of 3DTilesRendererJS and three-loader-3dtiles were not conclusive, so we decided to not use these library to support 3D Tiles and to continue having a native support of 3d tiles. Therefore, we will improve our support of this format in the coming months.

We haven't tested loaders.gl loaders yet and we will test it for implementing i3dm support in itowns and use it if it is appropriate. We might also do a similar study when 3d tiles next will be released.

**Comment by alexLuky** [comment]
@jailln @mgermerie 
Is there any news about this [issue](https://github.com/iTowns/itowns/issues/1859) and sphere Cesium b3dm support?

**Reply by jailln** [reply]
No, not yet, but I tested your dataset with 3DTilesRendererJS integrated into itowns and the issue is still there so that gives some insights about where it might come and where it might not... I'll let you know when I found something

**Reply by alexLuky** [reply]
Ok, thank you!
